pkg/challenge/response.go

```diff
@@ -29,23 +32,58 @@ func (o *OuterResponse) Open(privateEncryptionKey [32]byte) (*InnerResponse, err
 		return nil, fmt.Errorf("unmarshaling inner box: %w", err)
 	}
 
-	counterPartyPubKey, err := echelper.PublicPemToEcdsaKey(innerResponse.PublicSigningKey)
-	if err != nil {
-		return nil, fmt.Errorf("unmarshalling public ecdsa signing key from pem: %w", err)
+	if err := verifyWithKeyBytes(innerResponse.PublicSigningKey, innerResponseBytes, o.Sig); err != nil {
+		return nil, fmt.Errorf("verifying challenge signature: %w", err)
+	}
+
+	// no sig 2 provided, return what we have
+	if o.Sig2 == nil || len(o.Sig2) <= 0 {
+		// if there is no sig2, set public signing key 2 to nil just in case so that
+		// the consumer does not fasly assume it was used to perform a signature
+		innerResponse.PublicSigningKey2 = nil
+		return &innerResponse, nil
+	}
+
+	// got a signature 2, no public key 2
+	if innerResponse.PublicSigningKey2 == nil || len(innerResponse.PublicSigningKey2) <= 0 {
+		return nil, fmt.Errorf("have signature 2 but no public signing key 2")
 	}
 
-	if err := echelper.VerifySignature(*counterPartyPubKey, innerResponseBytes, o.Sig); err != nil {
-		return nil, fmt.Errorf("verifying challenge: %w", err)
+	if err := verifyWithKeyBytes(innerResponse.PublicSigningKey2, innerResponseBytes, o.Sig2); err != nil {
+		return nil, fmt.Errorf("verifying challenge signature 2: %w", err)
 	}
 
 	return &innerResponse, nil
 }
 
+func verifyWithKeyBytes(keyBytes []byte, msg []byte, sig []byte) error {
+	if strings.HasPrefix(string(keyBytes), "-----BEGIN PUBLIC KEY-----") {
+		key, err := echelper.PublicPemToEcdsaKey(keyBytes)
+		if err != nil {
+			return err
+		}
+		return echelper.VerifySignature(*key, msg, sig)
```

Comment:
Like the ruby, if feels a bit weird to have the two calls to `echelper.VerifySignature`

Reply:
cleaned this up